### PR TITLE
feat: add scenario tests and unit tests for GNS emission

### PR DESCRIPTION
**Pull request by jinoosss**
## Description

This PR enhances the test coverage for the GNS token emission system by adding comprehensive scenario tests and unit tests. The changes focus on improving the reliability and maintainability of the emission calculation logic and minting functionality.

### Key Changes

**Enhanced Unit Tests in `gns_test.gno`:**
- Added comprehensive test cases for `calculateAmountToMint` function covering various scenarios:
  - Single block minting for different halving years (1st, 2nd, 3rd year)
  - Multi-block minting calculations
  - Complete year emission calculations
  - Cross-year emission calculations
  - Edge cases including emission end height and pre-emission start height
  - Total emission calculation for all 12 years

- Added detailed test cases for `MintGns` function covering:
  - Successful minting by emission realm
  - Failure scenarios when system is halted
  - Permission validation for non-emission realm callers
  - Same block minting prevention

**Improved Halving Logic in `halving.gno`:**
- Enhanced `setAvgBlockTimeInMs` function to properly handle block time changes
- Added logic to recalculate remaining block counts when block time changes
- Improved end block height calculations for all halving years
- Added proper handling of current year vs future year calculations

### Test Coverage Improvements

The new tests provide comprehensive coverage for:
- Emission amount calculations across different time periods
- Halving year transitions and calculations
- Permission and access control validation
- System halt state handling
- Block time change scenarios
- Edge cases and boundary conditions

### Benefits

- **Reliability**: Comprehensive test coverage ensures emission calculations work correctly across all scenarios
- **Maintainability**: Well-structured test cases make it easier to verify changes and catch regressions
- **Documentation**: Test cases serve as living documentation of expected behavior
- **Quality Assurance**: Automated testing reduces the risk of bugs in critical emission functionality

### Technical Details

- All tests follow the existing testing patterns using `uassert` framework
- Test cases include detailed comments explaining expected values and calculations
- Proper setup and teardown procedures ensure test isolation
- Integration with halt system for comprehensive state testing

### CODE REVIEW

File: contract/r/gnoswap/gns/halving.gno [hunk 1]
```diff
@@ -339,18 +339,47 @@ func setAvgBlockTimeInMs(ms int64) {
...
+		// multiplying by previous block time (ms) and dividing by new block time (ms)
+		halvingStartHeight := hd.getStartBlockHeight(year)
+		halvingEndHeight := hd.getEndBlockHeight(year)
+
+		changedStartHeight := halvingStartHeight
+		if std.ChainHeight() > halvingStartHeight {
+			changedStartHeight = std.ChainHeight()
+		}
+
+		remainBlockNum := halvingEndHeight - halvingStartHeight
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply contract/r/gnoswap/gns/halving.gno:370]: I missed that. I think that's right.

```go
remainBlockNum := halvingEndHeight - changedStartHeight
```

**Review by notJoon — DISMISSED** [review]
LGTM

File: contract/r/gnoswap/gns/halving.gno [hunk 2]
```diff
@@ -339,18 +339,47 @@ func setAvgBlockTimeInMs(ms int64) {
...
+		if year < currentYear {
+			continue
+		}
+
 		maxAmt := hd.getMaxAmount(year)
 		// divide the annual issuance amount by the number of days
 		// and then by the number of blocks
 		dailyAmt := maxAmt / DAY_PER_YEAR
 		newAmtPerBlock := dailyAmt / blockPerDay
 		hd.setAmountPerBlock(year, newAmtPerBlock)
```
Comment (dongwon8247) [inline contract/r/gnoswap/gns/halving.gno:357]:
```suggestion
		maxAmt := hd.getMaxAmount(year)
		halvingEndHeight := hd.getEndBlockHeight(year)
		
		// Calculate remaining blocks with new block time
		remainBlockNum := halvingEndHeight - std.ChainHeight()
		changedRemainBlockNum := remainBlockNum * prevBlockTimeMs / ms
		
		// For current year, use remaining amount to avoid over/under emission
		if year == currentYear {
			mintedAmt := hd.getMintedAmount(year)
			leftAmt := maxAmt - mintedAmt
			
			if changedRemainBlockNum > 0 {
				newAmtPerBlock := leftAmt / changedRemainBlockNum
				hd.setAmountPerBlock(year, newAmtPerBlock)
			}
		} else {
			// For future years, use total year amount
			dailyAmt := maxAmt / DAY_PER_YEAR
			newAmtPerBlock := dailyAmt / blockPerDay
			hd.setAmountPerBlock(year, newAmtPerBlock)
		}
		
		changedEndHeight := std.ChainHeight() + changedRemainBlockNum
		hd.setEndBlockHeight(year, changedEndHeight)
```

File: contract/r/gnoswap/gns/halving.gno [hunk 3]
```diff
@@ -339,18 +339,47 @@ func setAvgBlockTimeInMs(ms int64) {
...
+
+		changedStartHeight := halvingStartHeight
+		if std.ChainHeight() > halvingStartHeight {
+			changedStartHeight = std.ChainHeight()
+		}
+
+		remainBlockNum := halvingEndHeight - changedStartHeight
+		changedRemainBlockNum := remainBlockNum * prevBlockTimeMs / ms
+
+		changedEndHeight := changedStartHeight + changedRemainBlockNum
```
Comment (dongwon8247) [inline contract/r/gnoswap/gns/halving.gno:373]:
```suggestion
```

this part is moved above, and can be removed

File: contract/r/gnoswap/gns/halving.gno [hunk 4]
```diff
@@ -339,18 +339,47 @@ func setAvgBlockTimeInMs(ms int64) {
...
+
+		hd.setEndBlockHeight(year, changedEndHeight)
```
Comment (dongwon8247) [inline contract/r/gnoswap/gns/halving.gno:375]:
```suggestion
```
this part is moved above, and can be removed

**Review by dongwon8247 — CHANGES_REQUESTED** [review]
@jinoosss Can you please add testcases with these scenarios?

1. Mid-Year Block Time Change Test
```
- Scenario: Change block time after 25%, 50%, 75% of a year
- Start Year 2 with 2s block time
- At 25% progress: Change to 4s
- Verify: 
  * Log emission rate before/after change
  * Total Year 2 emission = exactly 225T
  * Amount per block doubles from ~14M to ~28M
```
2. Current Year Remaining Amount Test
```
Scenario: Verify correct handling of already minted amount
- Progress to 40% of Year 3
- Log: Already minted amount
- Change block time 2s→3s
- Verify:
  * New amount per block = (remaining amount) / (remaining blocks)
  * NOT (total year amount) / (new blocks per year)
```
3. Extreme Ratio Changes
```
Scenario: Test 5x and 0.2x changes
- Test 1s→5s (5x increase)
- Test 5s→1s (0.2x decrease)
- Verify: No overflow/underflow, correct emission
```
4. Year Boundary Precise Check
```
Scenario: Verify no gaps between years after block time change
- Change block time mid-year
- Progress to year boundary
- Verify:
  * Year N ends at block X
  * Year N+1 starts at block X+1 (no gap)
  * Log exact transition blocks
```

File: contract/r/gnoswap/gns/halving.gno (hunk 2 above)
Comment (dongwon8247) [reply contract/r/gnoswap/gns/halving.gno:357]: For the current year, use the remaining amount to avoid over/under emission, and for future years, use the total year amount. This change is essential to prevent over/under emission by correctly calculating the per-block amount based on **remaining (not total)** emission for the current year when block time changes mid-year.

File: contract/r/gnoswap/gns/halving.gno [hunk 5]
```diff
@@ -339,18 +339,47 @@ func setAvgBlockTimeInMs(ms int64) {
...
+		// calculate changed end block height
+		// calculate new remaining blocks by taking original remaining blocks,
+		// multiplying by previous block time (ms) and dividing by new block time (ms)
+		halvingStartHeight := hd.getStartBlockHeight(year)
+		halvingEndHeight := hd.getEndBlockHeight(year)
+
+		changedStartHeight := halvingStartHeight
+		if std.ChainHeight() > halvingStartHeight {
+			changedStartHeight = std.ChainHeight()
+		}
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply contract/r/gnoswap/gns/halving.gno:368]: This code is essential for this feature.

This code sets the starting height for reapplying the block average time.

This code runs a loop based on the emission distribution year.
For the current year, it is set to the current block height instead of the previously set halving height.
For years after the current year, it is set based on the readjusted starting block height.

If write `halvingStartHeight := std.ChainHeight()` as suggested in the code, items after the current year will have an incorrect starting height.

File: contract/r/gnoswap/gns/halving.gno [hunk 6]
```diff
@@ -339,18 +339,47 @@ func setAvgBlockTimeInMs(ms int64) {
...
+
+		remainBlockNum := halvingEndHeight - halvingStartHeight
+		changedRemainBlockNum := remainBlockNum * prevBlockTimeMs / ms
+
+		changedEndHeight := changedStartHeight + changedRemainBlockNum
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply contract/r/gnoswap/gns/halving.gno:373]: It should be calculated based on the starting height that needs to be changed, not a fixed std.ChainHeight value.

[the remark this answers is not in the document]
Comment (dongwon8247) [reply contract/r/gnoswap/gns/halving.gno]: Hmm. The questions we should answer are: should year boundaries move when block time changes? do block time changes apply to future years? - yes to both, in my opinion, because the GNS emission schedule is time-based (12yr distribution, 2yr halving etc). Basically, you're saying each year is an "independent island", and I'm saying all years are a "continuous chain".

Let's compare with this scenario:
- Situation: the block time changes from 2s→4s at 25% of Year 2
- Initial State
	- Year 2: blocks `15,768,000` ~ `31,535,999`
	- Current: block `19,710,000` (25% progress)
	- Year 3: blocks `31,536,000` ~ `47,303,999` (scheduled)
- Calcylation for Year 3:
```
// Your approach
changedStartHeight = 31,536,000  // Year 3 original start
remainBlockNum = 47,303,999 - 31,536,000 = 15,767,999
changedRemainBlockNum = 15,767,999 * 2000/4000 = 7,883,999
changedEndHeight = 31,536,000 + 7,883,999 = 39,419,999

// My suggestion  
remainBlockNum = 47,303,999 - 19,710,000 = 27,593,999
changedRemainBlockNum = 27,593,999 * 2000/4000 = 13,796,999
changedEndHeight = 19,710,000 + 13,796,999 = 33,506,999
```
- The difference is `5,913,000` blocks (yours: `39,419,999`, mine: `39,419,999`)

What this means is, in your approach, even if Year 2 ends early, Year 3 tries to maintain its "original length", causing the total emission period to extend. This creates a problem where the emission period becomes way longer than 12 years.

File: contract/r/gnoswap/gns/halving.gno (hunk 2 above)
Comment (jinoosss) [reply contract/r/gnoswap/gns/halving.gno]: As you said, there was an error in the logic for updating the reward amount per block.

I believe this correction is correct.
I have corrected and simplified the logic.
https://github.com/gnoswap-labs/gnoswap/pull/686/commits/e29ebfd90e8c442f99ba317092b1b6ca4318f019

File: contract/r/gnoswap/gns/halving.gno (hunk 6 above)
[the remark this answers is not in the document]
Comment (dongwon8247) [reply contract/r/gnoswap/gns/halving.gno]: Check `mint_gns_by_change_block_time_filetest.gno` line 444-446: The test shows Year 2 end changes from `31,536,123` to `23,652,123` when block time changes, proving year boundaries should move.

**Comment by jinoosss** [comment]
> @jinoosss Can you please add testcases with these scenarios? I believe these are the essential ones we need to test it out before the production deployment.
> 
> 1. Mid-Year Block Time Change Test
> 
> ```
> - Scenario: Change block time after 25%, 50%, 75% of a year
> - Start Year 2 with 2s block time
> - At 25% progress: Change to 4s
> - Verify: 
>   * Log emission rate before/after change
>   * Total Year 2 emission = exactly 225T
>   * Amount per block doubles from ~14M to ~28M
> ```
> 
> 2. Current Year Remaining Amount Test
> 
> ```
> Scenario: Verify correct handling of already minted amount
> - Progress to 40% of Year 3
> - Log: Already minted amount
> - Change block time 2s→3s
> - Verify:
>   * New amount per block = (remaining amount) / (remaining blocks)
>   * NOT (total year amount) / (new blocks per year)
> ```
> 
> 3. Extreme Ratio Changes
> 
> ```
> Scenario: Test 5x and 0.2x changes
> - Test 1s→5s (5x increase)
> - Test 5s→1s (0.2x decrease)
> - Verify: No overflow/underflow, correct emission
> ```
> 
> 4. Year Boundary Precise Check
> 
> ```
> Scenario: Verify no gaps between years after block time change
> - Change block time mid-year
> - Progress to year boundary
> - Verify:
>   * Year N ends at block X
>   * Year N+1 starts at block X+1 (no gap)
>   * Log exact transition blocks
> ```

Thank you for writing the scenario. I will add this and ping you.

File: contract/r/gnoswap/gns/halving.gno (hunk 6 above)
[the remark this answers is not in the document]
Comment (dongwon8247) [reply contract/r/gnoswap/gns/halving.gno]: > One of the reasons for calculating independently is to ensure that the same number of blocks can be distributed for each year.

We should aim the same time duration (year, halving 2yr, total 12yr),not the same number of blocks.
